File: backend/app/schemas/movie.py
from typing import Optional, List
from pydantic import BaseModel, field_validator


TMDB_IMAGE_BASE = "https://image.tmdb.org/t/p/w500"


def normalize_poster_url(value: Optional[str]) -> Optional[str]:
    if not value:
        return None
    if value.startswith("//"):
        return f"https:{value}"
    if value.startswith("/"):
        return f"{TMDB_IMAGE_BASE}{value}"
    return value


class MovieBase(BaseModel):
    id: int
    title: str
    vote_average: float
    vote_count: int
    runtime: Optional[int] = None
    original_language: Optional[str] = None
    overview: Optional[str] = None
    popularity: float
    genres: Optional[str] = None
    keywords: Optional[str] = None
    year: Optional[int] = None
    poster_url: Optional[str] = None
    country: Optional[str] = None

    model_config = {"from_attributes": True}

    @field_validator("poster_url", mode="before")
    @classmethod
    def _normalize_poster_url(cls, value: Optional[str]) -> Optional[str]:
        return normalize_poster_url(value)


class MovieDetail(MovieBase):
    similar_movies: List[MovieBase] = []
    user_rating: Optional[float] = None
    clicks: int = 0
    views: int = 0


class MovieListItem(MovieBase):
    user_rating: Optional[float] = None


class MovieSearchResult(BaseModel):
    movies: List[MovieListItem]
    total: int
    page: int
    limit: int
    total_pages: int

File: frontend/src/types/movie.ts
export interface Movie {
  id: number;
  title: string;
  vote_average: number;
  vote_count: number;
  runtime: number | null;
  original_language: string | null;
  overview: string | null;
  popularity: number;
  genres: string | null;
  keywords: string | null;
  year: number | null;
  poster_url: string | null;
  country: string | null;
  user_rating?: number | null;
}

export interface MovieDetail extends Movie {
  similar_movies: Movie[];
  clicks: number;
  views: number;
}

export interface MovieSearchResult {
  movies: Movie[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

export interface MovieFilters {
  q?: string;
  genre?: string;
  year_from?: number;
  year_to?: number;
  language?: string;
  min_rating?: number;
  sort_by?: "popularity" | "rating" | "year" | "votes";
  page?: number;
  limit?: number;
}
